File: 코딩헬스장/15624. 피보나치 수 7/피보나치 수 7.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        System.out.println(fibonacci(n));
    }

    public static long fibonacci(int n) {
        if (n <= 0) {
            return 0;
        }
        long[] fib = new long[n+1];
        fib[0] = 0;
        fib[1] = 1;
        for (int i = 2; i <= n; i++) {
            fib[i] = (fib[i-1] + fib[i-2]) % 1000000007;
        }
        return fib[n];
    }
}

// 메모리, 시간 개선 코드
import java.util.Scanner;

public class Main {

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int n = scanner.nextInt();
    System.out.println(fibonacci(n)); 
  }

  public static long fibonacci(int n) {
    if (n < 0) return 0;
    
    long first = 0; 
    long second = 1;
    for (int i = 1; i <= n; i++) {
      long temp = (first + second) % 1000000007;
      first = second;
      second = temp;
    }
    return first;
  }

}

File: 코딩헬스장/15624. 피보나치 수 7/피보나치 수 7.kt
import java.util.Scanner

fun main(args: Array<String>) {
    val scanner = Scanner(System.`in`)
    val n = scanner.nextInt()
    println(fibonacci(n))
}

fun fibonacci(n: Int): Long {
    if (n <= 1) return n.toLong()
    var a = 0L
    var b = 1L
    for (i in 2..n) {
        val temp = b // 이전 값만 저장하기 위해 temp 변수 사용
        b = (a + b) % 1000000007 
        a = temp
    }
    return b
}